daemon/algod/api/server/v2/simulate.go

```diff
@@ -0,0 +1,205 @@
+func (l apiSimulatorLedgerConnector) GetCreatorForRound(round basics.Round, cidx basics.CreatableIndex, ctype basics.CreatableType) (creator basics.Address, ok bool, err error) {
+	if round != l.Latest() {
+		err = fmt.Errorf(
+			"GetCreatorForRound() evaluator called this function for the wrong round %d, "+
+				"latest round is %d",
+			round, l.Latest())
+		return
+	}
+
+	return l.GetCreator(cidx, ctype)
+}
+
+func makeLedgerForSimulatorFromLedgerForAPI(ledgerForAPI LedgerForAPI, hdr bookkeeping.BlockHeader) LedgerForSimulator {
+	return &apiSimulatorLedgerConnector{ledgerForAPI, hdr}
+}
+
+// ==============================
+// > Simulator Errors
+// ==============================
+
+// SimulatorError is the base error type for all simulator errors.
+type SimulatorError struct {
+	error
+}
+
+// InvalidTxGroupError occurs when an invalid transaction group was submitted to the simulator.
+type InvalidTxGroupError struct {
+	SimulatorError
+}
+
+// InvalidSignatureError occurs when a transaction has an invalid signature.
+type InvalidSignatureError struct {
+	SimulatorError
+}
+
+// ScopedSimulatorError is a simulator error that has 2 errors, one for internal use and one for
+// displaying publicly. THe external error is useful for API routes/etc.
+type ScopedSimulatorError struct {
+	SimulatorError        // the original error for internal use
+	External       string // the external error for public use
+}
+
+// ==============================
+// > Simulator Helper Methods
+// ==============================
+
+func isInvalidSignatureError(err error) bool {
+	var signatureError verify.SignatureError
+	if errors.As(err, &signatureError) {
+		return true
+	}
+
+	return false
+}
+
+// ==============================
+// > Simulator
+// ==============================
+
+// Simulator is a transaction group simulator for the block evaluator.
+type Simulator struct {
+	ledger LedgerForSimulator
+}
+
+// MakeSimulator creates a new simulator from a ledger.
+func MakeSimulator(ledger LedgerForSimulator) *Simulator {
+	return &Simulator{
+		ledger: ledger,
+	}
+}
+
+// MakeSimulatorFromAPILedger creates a new simulator from an API ledger.
+func MakeSimulatorFromAPILedger(ledgerForAPI LedgerForAPI, hdr bookkeeping.BlockHeader) *Simulator {
+	ledger := makeLedgerForSimulatorFromLedgerForAPI(ledgerForAPI, hdr)
+	return MakeSimulator(ledger)
+}
+
+// checkWellFormed checks that the transaction is well-formed. A failure message is returned if the transaction is not well-formed.
+func (s Simulator) checkWellFormed(txgroup []transactions.SignedTxn) error {
+	hdr, err := s.ledger.BlockHdr(s.ledger.Latest())
+	if err != nil {
+		return ScopedSimulatorError{SimulatorError{fmt.Errorf("please contact us, this shouldn't happen. Current block error: %v", err)}, "current block error"}
+	}
+
+	_, err = verify.TxnGroup(txgroup, hdr, nil)
+	if err != nil {
+		// invalid signature error
+		if isInvalidSignatureError(err) {
```

Comment:
@jdtzmn Request for clarification:  What motivates partitioning signature errors as a discrete error class?

I'm not necessarily opposed.  For now, I'm trying to better understand the intuition.

Reply:
I believe this was discussed during our live discussion yesterday. I'll briefly re-summarize just in case. Signature errors are distinct because `verify.TxnGroup` also calls `WellFormed` which does basic transaction validation and we'd like the simulate endpoint handler to respond to signature errors and invalid transaction errors differently.

Reply:
@jdtzmn Thanks for the summary, optionally:  Are you up for adding an inline comment near here explaining _why_  signature failures receive distinct treatment?

From yesterday's verbal, I took away that the motivator is end user convenience for rapid experimentation.  Though I don't think the rationale is clear from a source code inspection.

If you prefer as is or feel it's already addressed, feel welcomed to resolve.

Reply:
We did say this, but I just submitted a comment that potentially simplifies things from there.  I don't think we should allow the assortment of signature problems allowed by isIvalidSignatureError(). We should just allow specifically a missing signature.